elks/fs/romfs/romfs-low.S

```diff
@@ -0,0 +1,250 @@
+; ROMFS - A tiny filesystem in ROM
+
+; This module implements low-level functions
+; to directly access the filesystem in ROM,
+; and used by the ROMFS high-level module.
+
+#define __ASSEMBLY__
+#include <linuxmt/config.h>
+
+	.text
+
+; Get superblock
+; int romfs_super_get (byte_t *);
+; TODO: add argument with size of destination buffer
+
+_romfs_super_get:
+
+	push  bp
+	mov   bp,sp
+	push  ds
+	push  di
+	push  si
+	push  cx
+
+	mov   ax,#CONFIG_ROMFS_BASE
+	mov   ds,ax
+
+	; TODO: check super block magic
+
+	xor   si,si
+	mov   cx,[si+6]  ; size of superblock
+	mov   di,[bp+4]  ; arg0
+
+	cld
+	rep
+	movsb
+
+	xor   ax,ax
+
+	pop   cx
+	pop   si
+	pop   di
+	pop   ds
+	pop   bp
+	ret
+
+
+; Get inode from index
+; int romfs_inode_get (word_t index, byte_t *);
+; TODO: add argument with size of destination buffer
+
+_romfs_inode_get:
+
+	push  bp
+	mov   bp,sp
+	push  ds
+	push  di
+	push  si
+	push  dx
+	push  cx
+
+	mov   ax,#CONFIG_ROMFS_BASE
+	mov   ds,ax
+
+	; Check inode index
+
+	xor   si,si
+
+	mov   ax,[bp+4]   ; arg0
+	cmp   ax,[si+10]  ; inodes count
+	jb    inode_get_next1
+
+	mov   ax,#1
+	jmp   inode_get_exit
+
+inode_get_next1:
+
+	; Compute inode base from index
+
+	xor   dx,dx
+	mov   cx,[si+8]  ; inode size
+	mul   cx
```

Comment:
Is this aligned in some way that would allow a faster left bit shift to be used instead of a multiply?

Reply:
Well, one goal of the ROMFS is to take advantage of the flat nature of the ROM, to avoid wasting space with alignments like in other filesystems that divide space in blocks. I only align the files on paragraphs for future in-place execution, as suggested by Juan on the mailing list.

So yes, I agree MUL is a costly operation, but it helps to have a minimal size of the inode, and that cost is balanced by direct access to the filesystem in ROM. I am not sure padding the inode from 6 to 8 bytes and replacing MUL by SHL would improve the overall performance a lot, at the cost of reducing by 25% the available space for inodes in the first data segment (so limited to 64 KB).